### PR TITLE
Multiport

### PR DESCRIPTION
**Pull request by ndhanushkodi**
Based off of research and POC by @lkysow https://github.com/hashicorp/consul-k8s/pull/754 Thank you!

Note to reviewers:
I recommend reading the acceptance test first, then looking at the implementation. In the acceptance tests, I've changed the signature for `CheckStaticServerConnection*` functions, so there's a lot of 1 line changes all over the acceptance tests due to that.

Changes proposed in this PR:

Support a workaround for multi port pods by registering a Consul service per port, and by injecting init containers and envoy sidecars per port.

- Acceptance test for multiport
  - The multiport app is just http-echo listening on different ports in 2 containers
  - First, the test deploys `static-client` and `multiport`, with static-client having `multiport` and `multiport-admin` as upstreams. It checks that `static-client` can make connections to each service-- `multiport`, and `multiport-admin`. This is to test inbound connections for a multiport app.
  - Then the test deploys `static-server`, and checks that `multiport` can make a connection to `static-server`. This is to test outbound connections from a multiport app. Note that there is only 1 intention because all upstream connections are configured and go through the 1st service's envoy proxy.
  
- Implementation for multiport
  - `handler`
    - uses the service name annotation to figure out if this is a multiport pod or not
    - inject a connect-init container and an envoy-sidecar container per port/service
    - mount all service account tokens not already on the pod. For any service whose service account is not already attached to the pod, this mounts a volume for that service account token. In container-init, these volumes are added as a container volumeMount so the service account token file can be passed to the connect-init command via `-bearer-token-file`
  - `container_init` -- update the template to pass in arguments specific to the multiport service
  - `envoy_sidecar` -- allow the bootstrap file path to have the service name in it for multiport services, and add `--base-id`  since there could be multiple envoy sidecars
  - `endpoints_controller`: the listener port should be 20000 for the first service, 20001 for the second, etc.
  - `connect_init` command: when its a multiport app, add a filter to poll by service name, use the service specific bearertoken, acl-token-sink, and proxyid files
  
How I've tested this PR:
acceptance tests and unit tests

How I expect reviewers to test this PR:
see note above, and code review.

Checklist:
- [x] Tests added
- [ ] CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

### CODE REVIEW

File: acceptance/framework/k8s/deploy.go [hunk 1]
```diff
@@ -16,8 +16,6 @@ import (
 	"k8s.io/apimachinery/pkg/util/yaml"
 )
```
Comment (ndhanushkodi) [inline acceptance/framework/k8s/deploy.go:18]: Changed the function signature for CheckStaticServer* to allow passing in an `app` which is the app you're connecting from, and an `expectedSuccessOutput` which is the expected message back from the server. I needed to be able to make connections from the multiport app and not just static client, and each of the multiport services had a different success output.

File: acceptance/tests/connect/connect_helper.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ import (
 
 const staticClientName = "static-client"
```
Comment (ndhanushkodi) [inline acceptance/tests/connect/connect_helper.go:20]: This file is function signature changes

File: acceptance/tests/connect/connect_inject_namespaces_test.go [hunk 3]
```diff
@@ -177,9 +177,9 @@ func TestConnectInjectNamespaces(t *testing.T) {
 			if c.secure {
 				logger.Log(t, "checking that the connection is not successful because there's no intention")
 				if cfg.EnableTransparentProxy {
-					k8s.CheckStaticServerConnectionFailing(t, staticClientOpts, fmt.Sprintf("http://static-server.%s", staticServerNamespace))
```
Comment (ndhanushkodi) [inline acceptance/tests/connect/connect_inject_namespaces_test.go:180]: function signature changes

File: acceptance/tests/connect/connect_inject_test.go [hunk 4]
```diff
@@ -148,8 +150,169 @@ func TestConnectInject_RestartConsulClients(t *testing.T) {
...
 	logger.Log(t, "checking that connection is still successful")
 	if cfg.EnableTransparentProxy {
-		k8s.CheckStaticServerConnectionSuccessful(t, ctx.KubectlOptions(t), "http://static-server")
+		k8s.CheckStaticServerConnectionSuccessful(t, ctx.KubectlOptions(t), staticClientName, "http://static-server")
 	} else {
-		k8s.CheckStaticServerConnectionSuccessful(t, ctx.KubectlOptions(t), "http://localhost:1234")
+		k8s.CheckStaticServerConnectionSuccessful(t, ctx.KubectlOptions(t), staticClientName, "http://localhost:1234")
+	}
+}
+
```
Comment (ndhanushkodi) [inline acceptance/tests/connect/connect_inject_test.go:158]: Adds the multiport acceptance test

File: acceptance/tests/fixtures/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,2 @@
+FROM kschoche/http-echo:latest
```
Comment (ndhanushkodi) [inline acceptance/tests/fixtures/Dockerfile:1]: This adds curl to the http-echo image, I plan to swap out this for the actual http-echo image, so this can be ignored for now in review-- the only change would be to delete this file and swap the image.

File: acceptance/tests/fixtures/bases/multiport-app/anyuid-scc-rolebinding.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (ndhanushkodi) [inline acceptance/tests/fixtures/bases/multiport-app/anyuid-scc-rolebinding.yaml:1]: Added a new multiport app-- this was much easier to read than trying to patch the existing static-server app. It just runs http-echo on two different ports with 2 different messages.

File: acceptance/tests/ingress-gateway/ingress_gateway_namespaces_test.go [hunk 7]
```diff
@@ -128,7 +128,7 @@ func TestIngressGatewaySingleNamespace(t *testing.T) {
 				// via the bounce pod. It should fail to connect with the
```
Comment (ndhanushkodi) [inline acceptance/tests/ingress-gateway/ingress_gateway_namespaces_test.go:128]: function signature changes

File: acceptance/tests/ingress-gateway/ingress_gateway_test.go [hunk 8]
```diff
@@ -13,6 +13,8 @@ import (
 	"github.com/stretchr/testify/require"
 )
```
Comment (ndhanushkodi) [inline acceptance/tests/ingress-gateway/ingress_gateway_test.go:15]: Function signature changes

File: acceptance/tests/mesh-gateway/mesh_gateway_test.go [hunk 9]
```diff
@@ -127,7 +127,7 @@ func TestMeshGatewayDefault(t *testing.T) {
 	k8s.DeployKustomize(t, primaryContext.KubectlOptions(t), cfg.NoCleanupOnFailure, cfg.DebugDirectory, "../fixtures/cases/static-client-multi-dc")
 
 	logger.Log(t, "checking that connection is successful")
```
Comment (ndhanushkodi) [inline acceptance/tests/mesh-gateway/mesh_gateway_test.go:129]: Function signature changes

File: acceptance/tests/partitions/partitions_test.go [hunk 10]
```diff
@@ -415,11 +415,11 @@ func TestPartitions(t *testing.T) {
 				if c.ACLsAndAutoEncryptEnabled {
 					logger.Log(t, "checking that the connection is not successful because there's no intention")
 					if cfg.EnableTransparentProxy {
```
Comment (ndhanushkodi) [inline acceptance/tests/partitions/partitions_test.go:417]: Function signature changes

File: acceptance/tests/terminating-gateway/terminating_gateway_namespaces_test.go [hunk 11]
```diff
@@ -121,7 +121,7 @@ func TestTerminatingGatewaySingleNamespace(t *testing.T) {
```
Comment (ndhanushkodi) [inline acceptance/tests/terminating-gateway/terminating_gateway_namespaces_test.go:121]: Function signature changes

File: acceptance/tests/terminating-gateway/terminating_gateway_test.go [hunk 12]
```diff
@@ -93,7 +93,7 @@ func TestTerminatingGateway(t *testing.T) {
 
 			// Test that we can make a call to the terminating gateway.
```
Comment (ndhanushkodi) [inline acceptance/tests/terminating-gateway/terminating_gateway_test.go:94]: Function signature changes

File: acceptance/tests/vault/vault_test.go [hunk 13]
```diff
@@ -15,7 +15,8 @@ import (
 )
```
Comment (ndhanushkodi) [inline acceptance/tests/vault/vault_test.go:16]: Function signature changes

File: control-plane/connect-inject/endpoints_controller_test.go [hunk 14]
```diff
@@ -538,6 +538,383 @@ func TestProcessUpstreams(t *testing.T) {
...
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+
+			svcName := getServiceName(*tt.pod(), *tt.endpoint)
+			require.Equal(t, tt.expSvcName, svcName)
+
+		})
+	}
+}
+
```
Comment (ndhanushkodi) [inline control-plane/connect-inject/endpoints_controller_test.go:603]: Kept this TestReconcileCreateEndpoint_MultiportService test separate from the existing create endpoint test because of the changes to the case structure that wouldn't be needed by the non-multiport tests. Also because this is a workaround, it might be nice to keep the code as separate as possible.

**Review by ndhanushkodi — COMMENTED** [review]
Adding some comments for reviewers since theres a lot of files

File: control-plane/connect-inject/handler.go [hunk 15]
```diff
@@ -225,21 +231,92 @@ func (h *Handler) Handle(ctx context.Context, req admission.Request) admission.R
...
+	annotatedSvcNames := h.annotatedServiceNames(pod)
+	multiPort := len(annotatedSvcNames) > 1
 
-	// Add the Envoy sidecar.
-	envoySidecar, err := h.envoySidecar(*ns, pod)
-	if err != nil {
-		h.Log.Error(err, "error configuring injection sidecar container", "request name", req.Name)
-		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("error configuring injection sidecar container: %s", err))
+	// For single port pods, add the single init container and envoy sidecar.
+	if len(annotatedSvcNames) == 0 || len(annotatedSvcNames) == 1 {
```
Comment (ishustava) [inline control-plane/connect-inject/handler.go:240]: I think that's equivalent?
```suggestion
	if len(annotatedSvcNames) < 2 {
```

Comment (ishustava) [reply control-plane/connect-inject/handler.go:240]: or maybe this could be `if !multiPort` and then the following `if multiPort` could be an `else`?

**Review by ishustava — COMMENTED** [review]
This looks fantastic!!! Thanks you so much for comments throughout the code and the PR - this made it so easy to review ❤️ 

I left some comments and questions inline, mostly to clarify my own understanding and correct some minor things.

Great work on this!!

File: charts/consul/templates/connect-inject-clusterrole.yaml [hunk 16]
```diff
@@ -49,5 +49,10 @@ rules:
   - {{ template "consul.fullname" . }}-connect-inject-acl-token
   verbs:
   - get
+- apiGroups: [""]
+  resources:
+    - serviceaccounts
+  verbs:
+    - get
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply charts/consul/templates/connect-inject-clusterrole.yaml:56]: So since we're mounting the service account secret on the volume, the reason we're getting the service account in the handler is to [get the secret name](https://github.com/hashicorp/consul-k8s/pull/1012/files/9e7920439cc3b8f2e3aae447e06ef787dab2b480#diff-9edb67e94492a07a3800e8add0f97513e33e559038ec532ced787d9af7b63f09R284) (and to do validation that the service account exists). The secret name would be something like: `web-token-bvbk5` so we can't infer the whole name from the service account name, so I think we need this perm for that

[the remark this answers is not in the document]
Comment (ishustava) [reply charts/consul/templates/connect-inject-clusterrole.yaml:56]: Ah ok! that makes sense. I thought for a second that we may be able to just mount the secret but you're right we can't infer it. I guess the only question is whether we should put it behind a feature flag since we only need it for multiport? I think it's also OK to leave as is and change if someone raises an issue.

File: acceptance/framework/k8s/deploy.go [hunk 17]
```diff
@@ -77,10 +75,10 @@ func DeployKustomize(t *testing.T, options *k8s.KubectlOptions, noCleanupOnFailu
 // to be "hello world" in a case of success.
 // If expectSuccess is true, it will expect connection to succeed,
 // otherwise it will expect failure due to intentions.
-func CheckStaticServerConnection(t *testing.T, options *k8s.KubectlOptions, expectSuccess bool, failureMessages []string, curlArgs ...string) {
+func CheckStaticServerConnection(t *testing.T, options *k8s.KubectlOptions, app string, expectSuccess bool, failureMessages []string, expectedSuccessOutput string, curlArgs ...string) {
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply acceptance/framework/k8s/deploy.go]: yeah definitely good idea!

File: acceptance/framework/k8s/deploy.go [hunk 18]
```diff
@@ -77,10 +75,10 @@ func DeployKustomize(t *testing.T, options *k8s.KubectlOptions, noCleanupOnFailu
 // to be "hello world" in a case of success.
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply acceptance/framework/k8s/deploy.go]: yup, updated!

File: acceptance/tests/connect/connect_inject_test.go [hunk 19]
```diff
@@ -148,8 +150,169 @@ func TestConnectInject_RestartConsulClients(t *testing.T) {
...
+const multiport = "multiport"
+const multiportAdmin = "multiport-admin"
+
+// Test that Connect works in a default and a secure installation.
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply acceptance/tests/connect/connect_inject_test.go:162]: yup thanks for catching!

File: acceptance/tests/connect/connect_inject_test.go [hunk 20]
```diff
@@ -148,8 +150,169 @@ func TestConnectInject_RestartConsulClients(t *testing.T) {
...
+					Name: multiport,
+					Sources: []*api.SourceIntention{
+						{
+							Name:   staticClientName,
+							Action: api.IntentionActionAllow,
+						},
+					},
+				}, nil)
+				require.NoError(t, err)
+				logger.Log(t, fmt.Sprintf("creating intention for %s", staticServerAdminName))
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply acceptance/tests/connect/connect_inject_test.go:251]: yup!

File: acceptance/tests/connect/connect_inject_test.go [hunk 21]
```diff
@@ -148,8 +150,169 @@ func TestConnectInject_RestartConsulClients(t *testing.T) {
...
+	cases := []struct {
+		secure      bool
+		autoEncrypt bool
+	}{
+		{false, false},
+		{true, false},
+		{true, true},
+	}
+
+	for _, c := range cases {
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply acceptance/tests/connect/connect_inject_test.go:173]: Yeah, it was totally just running in the tproxy acceptance tests but using explicit upstreams-- added the skip, thanks!

File: control-plane/connect-inject/container_init.go [hunk 22]
```diff
@@ -142,12 +149,42 @@ func (h *Handler) containerInit(namespace corev1.Namespace, pod corev1.Pod) (cor
 		TProxyExcludeUIDs:          splitCommaSeparatedItemsFromAnnotation(annotationTProxyExcludeUIDs, pod),
 		ConsulDNSClusterIP:         consulDNSClusterIP,
 		EnvoyUID:                   envoyUserAndGroupID,
+		MultiPort:                  multiPort,
+		EnvoyAddressPort:           20000 + mpi.serviceIndex,
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/container_init.go:153]: I think you are right, it might be redundant from what we're doing in endpoint controller for the proxy service registration, I'll try removing this and running acceptance tests!

File: control-plane/connect-inject/endpoints_controller_test.go [hunk 23]
```diff
@@ -538,6 +538,383 @@ func TestProcessUpstreams(t *testing.T) {
...
+							DestinationServiceID:   "pod1-web-admin",
+							LocalServiceAddress:    "127.0.0.1",
+							LocalServicePort:       9090,
+							Upstreams: []api.Upstream{
+								{
+									DestinationType: api.UpstreamDestTypeService,
+									DestinationName: "upstream1",
+									LocalBindPort:   1234,
+								},
+							},
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/endpoints_controller_test.go:749]: you're totally right-- I think it was still working as expected with both proxies configuring the same set of upstreams, but its redundant and a weird edge case and not as we planned to implement in the RFC. Great catch!! I'll update the test and implementation

File: control-plane/connect-inject/envoy_sidecar.go [hunk 24]
```diff
@@ -10,19 +10,25 @@ import (
...
-func (h *Handler) envoySidecar(namespace corev1.Namespace, pod corev1.Pod) (corev1.Container, error) {
+func (h *Handler) envoySidecar(namespace corev1.Namespace, pod corev1.Pod, mpi multiPortInfo) (corev1.Container, error) {
 	resources, err := h.envoySidecarResources(pod)
 	if err != nil {
 		return corev1.Container{}, err
 	}
 
-	cmd, err := h.getContainerSidecarCommand(pod)
+	multiPort := mpi != multiPortInfo{}
+	cmd, err := h.getContainerSidecarCommand(pod, mpi.serviceName, mpi.serviceIndex)
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/envoy_sidecar.go:20]: yup, it checks whether mpi.ServiceName is empty to determine how to configure the command!

File: control-plane/connect-inject/endpoints_controller_test.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/endpoints_controller_test.go]: Updated the implementation and test!

File: acceptance/framework/k8s/deploy.go [hunk 25]
```diff
@@ -90,19 +88,24 @@ func CheckStaticServerConnection(t *testing.T, options *k8s.KubectlOptions, expe
 // If expectSuccess is true, it will expect connection to succeed,
 // otherwise it will expect failure due to intentions. If multiple failureMessages are provided it will assert
 // on the existence of any of them.
-func CheckStaticServerConnectionMultipleFailureMessages(t *testing.T, options *k8s.KubectlOptions, expectSuccess bool, failureMessages []string, curlArgs ...string) {
+func CheckStaticServerConnectionMultipleFailureMessages(t *testing.T, options *k8s.KubectlOptions, app string, expectSuccess bool, failureMessages []string, expectedSuccessOutput string, curlArgs ...string) {
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply acceptance/framework/k8s/deploy.go]: Nearly every call is setting "" and using the default, except for CheckStaticServerConnectionSuccessfulWithMessage, which is what we're using in the multiport test

File: charts/consul/templates/connect-inject-clusterrole.yaml (hunk 16 above)
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply charts/consul/templates/connect-inject-clusterrole.yaml:56]: Yeah, that sounds good, I'll go ahead and leave it then and see if it becomes an issue to allow this extra perm

File: control-plane/connect-inject/envoy_sidecar.go [hunk 26]
```diff
@@ -10,19 +10,25 @@ import (
...
 
-func (h *Handler) envoySidecar(namespace corev1.Namespace, pod corev1.Pod) (corev1.Container, error) {
+func (h *Handler) envoySidecar(namespace corev1.Namespace, pod corev1.Pod, mpi multiPortInfo) (corev1.Container, error) {
 	resources, err := h.envoySidecarResources(pod)
 	if err != nil {
 		return corev1.Container{}, err
 	}
 
-	cmd, err := h.getContainerSidecarCommand(pod)
+	multiPort := mpi != multiPortInfo{}
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/envoy_sidecar.go]: Updated!

File: control-plane/connect-inject/container_init.go [hunk 27]
```diff
@@ -72,6 +72,11 @@ type initContainerCommandData struct {
 
 	// ConsulDNSClusterIP is the IP of the Consul DNS Service.
 	ConsulDNSClusterIP string
+
+	MultiPort        bool
+	EnvoyAdminPort   int
+	EnvoyAddressPort int
+	BearerTokenFile  string
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/container_init.go:79]: Yup thanks!

File: control-plane/connect-inject/handler.go (hunk 15 above)
Comment (ndhanushkodi) [reply control-plane/connect-inject/handler.go:240]: made this an if !multiport + else!

File: acceptance/tests/fixtures/Dockerfile (hunk 5 above)
Comment (ndhanushkodi) [reply acceptance/tests/fixtures/Dockerfile]: Swapped and deleted

File: acceptance/tests/connect/connect_inject_test.go [hunk 28]
```diff
@@ -148,8 +150,169 @@ func TestConnectInject_RestartConsulClients(t *testing.T) {
...
+				t.Cleanup(func() {
+					retrier := &retry.Timer{Timeout: 5 * time.Minute, Wait: 1 * time.Second}
+					retry.RunWith(retrier, t, func(r *retry.R) {
+						tokens, _, err := consulClient.ACL().TokenList(nil)
+						require.NoError(r, err)
+						for _, token := range tokens {
+							require.NotContains(r, token.Description, multiport)
+							require.NotContains(r, token.Description, multiportAdmin)
+							require.NotContains(r, token.Description, staticClientName)
+							require.NotContains(r, token.Description, staticServerName)
```
[the remark this answers is not in the document]
Comment (ishustava) [reply acceptance/tests/connect/connect_inject_test.go:215]: Ah!!! that makes sense!

**Comment by ndhanushkodi** [comment]
WANFed tests for vault are failing but it might be due to my image? I'll try rebasing and using a new image for acceptance tests, but since everything else is passing and this doesn't touch vault this is ready for a second round of review

File: control-plane/connect-inject/endpoints_controller.go [hunk 29]
```diff
@@ -1072,3 +1093,12 @@ func consulTags(pod corev1.Pod) []string {
...
 	return interpolatedTags
 }
+
+func getMultiPortIdx(pod corev1.Pod, serviceEndpoints corev1.Endpoints) int {
+	for i, name := range strings.Split(pod.Annotations[annotationService], ",") {
+		if name == getServiceName(pod, serviceEndpoints) {
+			return i
+		}
+	}
+	return -1
```
Comment (ishustava) [inline control-plane/connect-inject/endpoints_controller.go:1103]: nit: I think we could also return an error here (it feels a bit more idiomatic to go) but definitely not a blocker!

**Review by ishustava — APPROVED** [review]
🎉 🎉 🎉 

A couple of minor comments, but they're not blocking!

File: control-plane/subcommand/connect-init/command.go [hunk 30]
```diff
@@ -66,21 +67,16 @@ func (c *Command) init() {
 	c.flagSet.StringVar(&c.flagConsulServiceNamespace, "consul-service-namespace", "", "Consul destination namespace of the service.")
 	c.flagSet.StringVar(&c.flagServiceAccountName, "service-account-name", "", "Service account name on the pod.")
 	c.flagSet.StringVar(&c.flagServiceName, "service-name", "", "Service name as specified via the pod annotation.")
+	c.flagSet.StringVar(&c.flagBearerTokenFile, "bearer-token-file", defaultBearerTokenFile, "Path to service account token file.")
+	c.flagSet.StringVar(&c.flagACLTokenSink, "acl-token-sink", defaultTokenSinkFile, "Service name as specified via the pod annotation.")
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/subcommand/connect-init/command.go:71]: omg good catch

File: control-plane/connect-inject/envoy_sidecar.go [hunk 31]
```diff
@@ -62,7 +68,7 @@ func (h *Handler) envoySidecar(namespace corev1.Namespace, pod corev1.Pod) (core
 		// has only injected init containers so all containers defined in pod.Spec.Containers are from the user.
 		for _, c := range pod.Spec.Containers {
 			// User container and Envoy container cannot have the same UID.
-			if c.SecurityContext != nil && c.SecurityContext.RunAsUser != nil && *c.SecurityContext.RunAsUser == envoyUserAndGroupID {
+			if c.SecurityContext != nil && c.SecurityContext.RunAsUser != nil && *c.SecurityContext.RunAsUser == envoyUserAndGroupID && c.Image != h.ImageEnvoy {
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/envoy_sidecar.go:71]: When we're running multiple sidecars in the pod, they clash with the same uid/gid so this check allows using the same uid/gid for 2 envoy sidecars. Since tproxy isn't supported it seems ok to let them run with the same uid/gid.
